src/hotspot/share/gc/shared/referenceProcessor.cpp

```diff
@@ -771,16 +774,33 @@ void ReferenceProcessor::balance_queues(DiscoveredList ref_lists[])
 #endif
 }
 
-void ReferenceProcessor::process_soft_ref_reconsider(BoolObjectClosure* is_alive,
-                                                     OopClosure* keep_alive,
-                                                     VoidClosure* complete_gc,
-                                                     AbstractRefProcTaskExecutor* task_executor,
-                                                     ReferenceProcessorPhaseTimes* phase_times) {
-  assert(!processing_is_mt() || task_executor != NULL, "Task executor must not be NULL when mt processing is set.");
+void ReferenceProcessor::run_task(RefProcTask& task, RefProcProxyTask& proxy_task, bool marks_oops_alive) {
+  WorkGang* gang = Universe::heap()->safepoint_workers();
+  assert(gang != NULL || !processing_is_mt(), "can not dispatch multi threaded without a work gang");
+  log_debug(gc, ref)("ReferenceProcessor::execute queues: %d, %s, marks_oops_alive: %s",
+                     num_queues(),
+                     processing_is_mt() ? "RefProcThreadModel::Multi" : "RefProcThreadModel::Single",
+                     marks_oops_alive ? "true" : "false");
+
+  proxy_task.prepare_run_task(task, num_queues(), processing_is_mt() ? RefProcThreadModel::Multi : RefProcThreadModel::Single, marks_oops_alive);
+  if (gang != NULL && processing_is_mt()) {
```

Comment:
Since `gang` is only used in thin branch, moving it inside `if` might be better. Besides, the assertion on L779 could be replaced by the code below, IMO.
```
if (processing_is_mt()) {
  WorkGang* gang = Universe::heap()->safepoint_workers();
  assert(gang != nullptr, "...");
  ...
}
```

Reply:
Yes!